pkg/controller/jobs/kubeflow/trainjob/trainjob_controller.go

```diff
@@ -0,0 +1,249 @@
+}
+
+func (t *TrainJob) IsActive() bool {
+	for i := range t.Status.JobsStatus {
+		if t.Status.JobsStatus[i].Active > 0 {
+			return true
+		}
+	}
+	return false
+}
+
+func (t *TrainJob) Suspend() {
+	t.Spec.Suspend = ptr.To(true)
+}
+
+func (t *TrainJob) Unsuspend() {
+	t.Spec.Suspend = ptr.To(false)
+}
+
+func (t *TrainJob) GVK() schema.GroupVersionKind {
+	return gvk
+}
+
+func (t *TrainJob) PodLabelSelector() string {
+	return fmt.Sprintf("%s=%s", jobsetapi.JobSetNameKey, t.Name)
+}
+
+func getChildJobSet(t *TrainJob) (*jobsetapi.JobSet, error) {
+	jobSet := jobsetapi.JobSet{}
+	// TODO: Ideally we should be using the parent context here
+	err := reconciler.client.Get(context.Background(), types.NamespacedName{Name: t.Name, Namespace: t.Namespace}, &jobSet)
+	return &jobSet, err
+}
+
+func (t *TrainJob) PodSets() ([]kueue.PodSet, error) {
+	jobset, err := getChildJobSet(t)
+	if err != nil {
+		return nil, err
+	}
+	return (*kJobset.JobSet)(jobset).PodSets()
+}
+
+func (t *TrainJob) RunWithPodSetsInfo(podSetsInfo []podset.PodSetInfo) error {
+	jobset, err := getChildJobSet(t)
+	if err != nil {
+		return err
+	}
+
+	if len(podSetsInfo) != len(jobset.Spec.ReplicatedJobs) {
+		return podset.BadPodSetsInfoLenError(len(jobset.Spec.ReplicatedJobs), len(podSetsInfo))
+	}
+
+	//TODO: Currently we can't restore the original TrainJob .Spec.PodSpecOverrides at job suspension, because the controller's admission webhook
+	// rejects any change to this property if there's any job active (which is very likely at that moment). Until we find a workaround, we always overwrite
+	// it at job start time. This will discard user provided .Spec.PodSpecOverrides.
+	t.Spec.PodSpecOverrides = []kftrainerapi.PodSpecOverride{}
+	for _, info := range podSetsInfo {
+		// The trainjob controller merges each podSpecOverride sequentially, so any existing user provided override will be processed first
+		t.Spec.PodSpecOverrides = append(t.Spec.PodSpecOverrides, kftrainerapi.PodSpecOverride{
+			TargetJobs: []kftrainerapi.PodSpecOverrideTargetJob{
+				{Name: string(info.Name)},
+			},
+			// TODO: Set the labels/annotations when supported. See https://github.com/kubeflow/trainer/pull/2785
+			NodeSelector:    info.NodeSelector,
+			Tolerations:     info.Tolerations,
+			SchedulingGates: info.SchedulingGates,
+		})
+	}
+	// Update the podSpecOverrides while the job is suspended, since is a requirement from the trainjob admission webhook
+	// TODO: Ideally we should be using the parent context here
+	err = reconciler.client.Update(context.Background(), t.Object())
+	if err != nil {
+		t.Spec.PodSpecOverrides = nil
+		return err
+	}
+
+	t.Unsuspend()
+	return nil
+}
+
+func (t *TrainJob) RestorePodSetsInfo(podSetsInfo []podset.PodSetInfo) bool {
+	// TODO: At this point, we should restore the .Spec.PodSpecOverrides, but the trainjob controller admission webhook
+	// rejects the change if any replicated pod is still active (which is very likely the case when restoring).
```

Comment:
This is very surprising, can you provide a reference? 

I would expect Kueue awaits for the suspended job to not be active first due to [this line](https://github.com/kubernetes-sigs/kueue/blob/3c3266be2f42cf216cb3e38a4fc863142e054529/pkg/controller/jobframework/reconciler.go#L535). 

Is it just a temporary failed request, or the integration does not work at all?

I would prefer all integrations follow the same pattern, so would like to first understand why we need this hack, and what makes the Trainer special.

Reply:
The admission webhook of the TrainJob controller doesn't allow changes to the `PodSpecOverrides` property when any job is active. This breaks workload evictions, since the job will never be stopped because there's nothing stopping it..

As discussed offline, I've implemented the `Stop` function to handle this situation and only consider the job stopped when there are no active jobs